### PR TITLE
IETF API-Problem specification for error messages

### PR DESCRIPTION
**Pull request by e0ipso**
Implements #27.

### CODE REVIEW

**Comment by amitaibu** [comment]
@cpliakas - how cool is that? :)

**Comment by e0ipso** [comment]
Tests fixed.

File: exceptions/RestfulException.php [hunk 1]
```diff
@@ -7,7 +7,91 @@
...
+  /**
+   * Return a string to the common problem type.
+   *
+   * @return string
+   *   URL pointing to the specific RFC-2616 section.
+   */
+  public function getType() {
+    // Depending on the error code we'll return a different URL.
+    $url = 'http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html';
+    $sections = array(
```
Comment (amitaibu) [inline exceptions/RestfulException.php:36]: You are crazy! ;)

File: exceptions/RestfulException.php [hunk 2]
```diff
@@ -7,7 +7,91 @@
...
+   *
+   * @return string
+   *   The url or NULL if empty.
+   */
+  public function getInstance() {
+    // Handle all instances using the advanced help module.
+    if (!module_exists('advanced_help') || empty($this->instance)) {
+      return;
+    }
+    return url($this->instance, array(
```
Comment (amitaibu) [inline exceptions/RestfulException.php:92]: minor, but I think it's more readable if it's in one line (I can change myself before merge).

File: exceptions/RestfulFloodException.php [hunk 3]
```diff
@@ -20,4 +20,12 @@ class RestfulFloodException extends RestfulException {
    * @var string
    */
   protected $description = 'Too Many Requests.';
+
+  /**
+   * Defines the problem instance.
+   *
+   * @var string
+   */
+  protected $instance = 'help/restful/problem-instances-flood';
```
Comment (amitaibu) [inline exceptions/RestfulFloodException.php:29]: maybe call it `$errorDescriptionPath`?

Comment (amitaibu) [inline exceptions/RestfulFloodException.php:29]: never mind, it's good enough.

File: tests/RestfulExceptionHandleTestCase.test [hunk 4]
```diff
@@ -29,11 +29,14 @@ class RestfulExceptionHandleTestCase extends DrupalWebTestCase {
...
     $this->drupalGet('api/v1/articles', $options);
     $expected_result = array(
-      'code' => 400,
-      'message' => 'The sort wrong_key is not allowed for this path.',
-      'description' => 'Bad Request.',
+      'type' => 'http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1',
+      'title' => 'The sort wrong_key is not allowed for this path.',
+      'status' => 400,
+      'detail' => 'Bad Request.',
+      'instance' => NULL,
```
Comment (amitaibu) [inline tests/RestfulExceptionHandleTestCase.test:36]: We could download advanced_help via .travis, so we could test the result here.

Comment (e0ipso) [inline tests/RestfulExceptionHandleTestCase.test:36]: Not sure it's worth the trouble.

File: exceptions/RestfulException.php (hunk 1 above)
Comment (e0ipso) [inline exceptions/RestfulException.php:36]: Some HTML c&p and text editor magic did the trick, and it adheres to the proposal! Also, bonus for linking to `W3.org`.

**Comment by amitaibu** [comment]
Code looks good, I think before merge it's worth adding a line or two in the README - care to do it?

**Comment by e0ipso** [comment]
Right! The section about Error Hnadling needs to be updated.

On Wed, May 28, 2014 at 8:13 PM, Amitai Burstein
notifications@github.comwrote:

> Code looks good, I think before merge it's worth adding a line or two in
> the README - care to do it?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Gizra/restful/pull/49#issuecomment-44444384
> .

## 

Mateu Aguiló Bosch.
Developer.

w. lb.cm/e0ipso
e. mateu@lullabot.com
skype. mateu.aguilo.bosch
irc. e0ipso

**Comment by e0ipso** [comment]
Added README notes.

**Comment by e0ipso** [comment]
This should be good to merge.

**Comment by amitaibu** [comment]
Merged, thanks!

**Comment by cpliakas** [comment]
@mateu-aguilo-bosch Great work!

@amitaibu, thanks for being open to this change and committing!

**Pull request closed without merging**